frame/support/src/traits/hooks.rs

```diff
@@ -82,11 +82,17 @@ pub trait OnIdle<BlockNumber> {
 #[impl_for_tuples(30)]
 impl<BlockNumber: Clone> OnIdle<BlockNumber> for Tuple {
 	fn on_idle(n: BlockNumber, remaining_weight: crate::weights::Weight) -> crate::weights::Weight {
-		let mut weight = 0;
+		let mut on_idle_functions = sp_std::vec::Vec::<fn(BlockNumber, crate::weights::Weight) -> crate::weights::Weight>::new();
 		for_tuples!( #(
-			let adjusted_remaining_weight = remaining_weight.saturating_sub(weight);
-			weight = weight.saturating_add(Tuple::on_idle(n.clone(),  adjusted_remaining_weight));
+			on_idle_functions.push(Tuple::on_idle);
 		)* );
+		let mut weight = 0;
+		for pallet_index in 0..on_idle_functions.len() {
```

Comment:
It would be great to start with `n % len` and then iterate all the modules so that a different module is processed first on every block (so that a heavy module cannot block the others).

Reply:
Just want to clarify.
Do you mean that first execution `on_idle` will be for the module which has index equal to `n % len` ? So the rest modules will be executed in the original order (excluding already executed module) ?
And if so, why is it preferable instead of the round robin execution for the every module on each new block ?

Reply:
Yes, the first execution in block `n` would be item `n % len`. The rest of the execution would be unchanged (you just start at a different position). This is preferable because one module might do a lot of processing in `on_idle` and "starve out" the others if it always executes before the later ones.

Reply:
But with the round robin execution we will have the same effect. For example we have 3 modules: `Module1`, `Module2`, `Module3`.
With the round robin algo which I am suggesting we will have the following behaviour:
1. `n = 0`, execution order  `Module1`, `Module2`, `Module3`
2. `n = 1`, execution order  `Module2`, `Module3`, `Module1`
3. `n = 2`, execution order  `Module3`, `Module1`, `Module2`
4. `n = 3`, execution order  `Module1`, `Module2`, `Module3`
As I understand you correct with the `n % len` execution we will have the following behaviour:
1. `n = 0`, execution order  `Module1`, `Module2`, `Module3`
2. `n = 1`, execution order  `Module2`, `Module1`, `Module3`
3.  `n = 2`, execution order  `Module3`, `Module1`, `Module2`
4. `n = 3`, execution order  `Module1`, `Module2`, `Module3`

In my opinion with the  `n % len` we have a one disadvantage, that for example `Module1` have a most privilege, with the round robin algo, which I am suggesting, every module will have the equal opportunity.

So what approach would we choose ?

Reply:
Oh, sorry, didn't see that you were already doing offsets based on the blocknumber below. All is well and never mind, then.